Adds possibility to add sprintf-values

heiglandreas opened the pull request:
This PR adds the possibility to use sprintf-formatting in the formatter strings.

Thus it's possible to mix intl-formatting with sprintf-formatting. So it's possible to do something like this:

``` php
echo $formatter->format('those %02d %s cost %number', 2, 'monkeys', 10023.23);
// de_DE: those 02 monkeys cost 10.023,23
// en_US: those 02 monkeys cost 10,023.23
```

heiglandreas committed:
Adds possibility to add sprintf-values

This commit adds the possibility to use sprintf-formatting in the
formatter strings.

SenseException commented:
Thank you for your time and effort you put into this PR. I haven't expect one for this young project. 

Because of the nature for internationalization, I'm not sure about the default `sprintf` type specifier. Shouldn't %d be formatted for the given locale, being an alias for %number?
IntlFormatter doesn't currently support all sprintf-known type specifier formats in `SprintfParser`.

SenseException commented on src/Formatter/SprintfFormatter.php:
```diff
@@ -0,0 +1,28 @@
+use DateTimeInterface;
+use DateTimeZone;
+use Budgegeria\IntlFormat\Exception\InvalidValueException;
+
```
All the use statements aren't used in this class.

SenseException commented on tests/Formatter/SprintfFormatterTest.php:
```diff
@@ -0,0 +1,37 @@
+     */
+    public function provideTypeSpecifier()
+    {
+        return [
```
A full list of type specifier would be nice.

heiglandreas committed:
Removes use-statements and adds more tests

heiglandreas commented:
> Because of the nature for internationalization, I'm not sure about the default sprintf type specifier. 

What default type? There is no default type for sprintf-values, or am I missing something?

> Shouldn't %d be formatted for the given locale, being an alias for %number?

`%d` formats integers according to the sprintf-rules and not to intl-rules. I wouldn't use it in this library in favour of `%number`  but as it's a valid value for `sprintf` I've added it. And I wouldn't use it as an alias to `%number` as that might have unintended side effects. IMO users should be able to decide whether to use `%d` or `%number`.

> IntlFormatter doesn't currently support all sprintf-known type specifier formats in SprintfParser. 

That should be the case. At least all type-specifiers described in http://php.net/sprintf are supported. Now also tested :wink: But you are right that not all options are tested on all types. I've added them selectively to test that they work in general. When edge-cases are reported the tests can then be extended with theses.

SenseException commented:
After some offline talk with @heiglandreas about this PR last weekend, I'll take a closer look on the regex as a last step before the merge.

SenseException commented:
@heiglandreas Okay, looks good to me. Could you please rebase to the current master?

heiglandreas commented:
Rebased to current master and ready to merge ;)

SenseException commented:
Merged. Thank you for your contribution.
